### PR TITLE
chore: support GSQL in tpcc benchmarking

### PR DESCRIPTION
**Pull request by hengfengli**
Currently, the data loader only supports to load data into a PG database. This PR adds:

* Support the GSQL dialect in the data loader. 
* Support the GSQL benchmark runner.

### CODE REVIEW

File: benchmarks/tpcc/README.md [hunk 1]
```diff
@@ -22,6 +22,27 @@ mvn spring-boot:run -Dspring-boot.run.arguments="
...
+  --tpcc.use-read-only-transactions=false
+  --tpcc.lock-scanned-ranges=false
+  --spanner.project=my-project
+  --spanner.instance=my-instance
+  --spanner.database=my-database
+  --pgadapter.credentials=/path/to/credentials.json
+  "
+```
+
+Load data into a GoogleSQL-dialect DB (we need to specify `client_lib_gsql` in `benchmark-runner` to indicate to load GSQL data):
```
Comment (olavloite) [inline benchmarks/tpcc/README.md]:
```suggestion
To load data into a GoogleSQL-dialect DB, you need to specify `client_lib_gsql` as the `benchmark-runner`. This instructs the benchmark application that it should use the Spanner Java client to connect to the database, and that it should use the GoogleSQL dialect:
```

File: benchmarks/tpcc/README.md [hunk 2]
```diff
@@ -41,7 +63,7 @@ mvn spring-boot:run -Dspring-boot.run.arguments="
 
 ### Run benchmark
 
-Currently, we support benchmark runners: `pgadapter`, `spanner_jdbc`, and `client_lib_pg`.
+Currently, we support benchmark runners: `pgadapter`, `spanner_jdbc`, `client_lib_pg`, and `client_lib_gsql`.
```
Comment (olavloite) [inline benchmarks/tpcc/README.md]:
```suggestion
Currently, we support the following benchmark runners: `pgadapter`, `spanner_jdbc`, `client_lib_pg`, and `client_lib_gsql`.
```

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/AbstractBenchmarkRunner.java [hunk 3]
```diff
@@ -143,6 +147,10 @@ private long getOtherWarehouseId(long currentId) {
     }
   }
 
+  private String getNowFunctionInQuery() {
+    return dialect == Dialect.POSTGRESQL ? "Now()" : "CURRENT_TIMESTAMP()";
```
Comment (olavloite) [inline benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/AbstractBenchmarkRunner.java]: You can replace this with `current_timestamp` (note: no parentheses), which works for both dialects.

GoogleSQL: https://cloud.google.com/spanner/docs/reference/standard-sql/timestamp_functions#current_timestamp
PostgreSQL: https://www.postgresql.org/docs/current/functions-datetime.html

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/SchemaService.java [hunk 4]
```diff
@@ -36,6 +36,8 @@ class SchemaService {
   }
 
   void createSchema() throws IOException, SQLException {
+    boolean is_googlesql = connectionUrl.startsWith("jdbc:cloudspanner");
```
Comment (olavloite) [inline benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/SchemaService.java]: Is this always correct? Don't we also support running the benchmark using the Spanner JDBC driver on a PostgreSQL-dialect database? Because in the latter case, the prefix of the connection URL will also be `jdbc:cloudspanner`.

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/SchemaService.java [hunk 5]
```diff
@@ -44,21 +46,24 @@ void createSchema() throws IOException, SQLException {
               .executeQuery(
                   "select count(1) "
                       + "from information_schema.tables "
-                      + "where table_schema='public' "
-                      + "and table_name in ('warehouse', 'district', 'customer', 'history', 'orders', 'new_orders', 'order_line', 'stock', 'item')")) {
+                      + "where "
+                      + (is_googlesql ? "" : "table_schema='public' and ")
```
Comment (olavloite) [inline benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/SchemaService.java]: This could break once we introduce named schema support.

```suggestion
                      + (is_googlesql ? "table_schema='' and " : "table_schema='public' and ")
```

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/dataloader/DataLoader.java [hunk 6]
```diff
@@ -52,6 +63,7 @@ public class DataLoader implements AutoCloseable {
   public DataLoader(
       DataLoadStatus status, String connectionUrl, TpccConfiguration tpccConfiguration) {
     this.connectionUrl = connectionUrl;
+    this.isGoogleSQL = connectionUrl.startsWith("jdbc:cloudspanner");
```
Comment (olavloite) [inline benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/dataloader/DataLoader.java]: Here also: I think this gives the wrong result if you are running a benchmark for a PostgreSQL-dialect database using the Spanner JDBC driver.

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/dataloader/DataLoader.java [hunk 7]
```diff
@@ -270,7 +282,125 @@ public long loadData() throws Exception {
...
+    long count = 0;
+    try (Connection connection = createConnection();
+        Statement statement = connection.createStatement()) {
+      // Unwrap the Cloud Spanner specific interface to be able to access custom
+      // methods.
+      CloudSpannerJdbcConnection spannerConnection =
+          connection.unwrap(CloudSpannerJdbcConnection.class);
+      spannerConnection.setAutoCommit(false);
+
+      List<Mutation> mutations = new ArrayList<>(Math.round(MUTATION_LIMIT_PER_COMMIT / 4));
```
Comment (olavloite) [inline benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/dataloader/DataLoader.java:354]: nit: Can you add a comment for why you do a 'divide limit by 4' here?

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/dataloader/DataLoader.java [hunk 8]
```diff
@@ -284,48 +414,67 @@ long loadTable(AbstractRowProducer rowProducer) throws SQLException, IOException
...
     // Allow copy operations to be non-atomic.
-    connection
-        .createStatement()
-        .execute("set spanner.autocommit_dml_mode='partitioned_non_atomic'");
+    if (isGoogleSQL) {
+      connection.createStatement().execute("set autocommit_dml_mode='partitioned_non_atomic'");
+    } else {
+      // Use upsert instead of insert for COPY to prevent data loading errors if the
+      // tables are
+      // already half-filled.
```
Comment (olavloite) [inline benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/dataloader/DataLoader.java]: nit: fix formatting

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/dataloader/OrderRowProducer.java [hunk 9]
```diff
@@ -13,16 +13,22 @@
...
 import java.util.Collections;
+import java.util.List;
 import java.util.Objects;
+import java.util.Random;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentMap;
 import java.util.stream.LongStream;
 
 class OrderRowProducer extends AbstractOrderedIdRowProducer {
+  private final long SEED = 12345;
```
Comment (olavloite) [inline benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/dataloader/OrderRowProducer.java]:
```suggestion
  private static final long SEED = 12345;
```

Comment (hengfengli) [reply benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/dataloader/OrderRowProducer.java]: Fixed.

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/dataloader/DataLoader.java (hunk 8 above)
Comment (hengfengli) [reply benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/dataloader/DataLoader.java]: Fixed.

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/dataloader/DataLoader.java (hunk 7 above)
Comment (hengfengli) [reply benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/dataloader/DataLoader.java:354]: Added a comment.

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/AbstractBenchmarkRunner.java (hunk 3 above)
Comment (hengfengli) [reply benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/AbstractBenchmarkRunner.java]: Updated.

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/SchemaService.java (hunk 5 above)
Comment (hengfengli) [reply benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/SchemaService.java]: Good catch. Fixed.

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/SchemaService.java (hunk 4 above)
Comment (hengfengli) [reply benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/SchemaService.java]: Good catch. Fixed.

File: benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/dataloader/DataLoader.java (hunk 6 above)
Comment (hengfengli) [reply benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/dataloader/DataLoader.java]: Updated as well.

**Review by olavloite — APPROVED** [review]
